framework/src/messaging/messaging_recv.c

```diff
@@ -0,0 +1,296 @@
+	}
+
+	info = (messaging_recv_info_t *)data->si_value.sival_ptr;
+	if (info == NULL) {
+		msgdbg("message recv fail : wrong param for wrapper callback(info).\n");
+		return;
+	}
+
+	while (1) {
+		ret = mq_receive(info->mqdes, (char *)&msg, sizeof(messaging_recv_info_t), 0);
+		if (ret < 0) {
+			msgdbg("message recv fail : mq_receive fail. %d\n", errno);
+			mq_close(info->mqdes);
+			MSG_FREE(data);
+			return;
+		}
+		/* Call user callback */
+		(*info->user_cb)(info->cb_data, &msg);
+
+		ret = mq_getattr(info->mqdes, &attr);
+		if (ret < 0) {
+			msgdbg("message recv fail : mq_getattr fail.\n");
+			mq_close(info->mqdes);
+			MSG_FREE(data);
+			return;
+		} else if (attr.mq_curmsgs == 0) {
+			/* All requests are handled. Register notification again. */
+			set_notification(info);
+			break;
+		}
+	}
+}
+
+/****************************************************************************
+ * Name : set_notify_signal
+ * 
+ * Description:
+ *  This function registers the wrapper callback for notification.
+ *
+ * Return Value:
+ *  On success, 0 (OK) is returned.; On failure, -1 (ERROR) is returned.
+ ****************************************************************************/
+static int set_notify_signal(void)
+{
+	int ret;
+	struct sigaction act;
+	act.sa_sigaction = (_sa_sigaction_t)wrapper_callback;
+	act.sa_flags = 0;
+	(void)sigemptyset(&act.sa_mask);
+
+	ret = sigaddset(&act.sa_mask, SIGMSG_UNICAST);
+	if (ret < 0) {
+		msgdbg("message recv fail : sigaddset error.\n");
+		return ERROR;
+	}
+
+	ret = sigaction(SIGMSG_UNICAST, &act, NULL);
+	if (ret == (int)SIG_ERR) {
+		msgdbg("message recv fail : sigaction error.\n");
+		return ERROR;
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name : messaging_recv_nonblock
+ * 
+ * Description:
+ *  This function is for non-blocking mq_receive.
+ * If there is a message in the queue, this function calls the callback with msg.
+ * If there is no message, then calls the notification function to register callback. 
+ *
+ * Input Parameters:
+ *  mqdes     : Message queue descriptor
+ *  port_name : The message port name to receive
+ *  msg       : The message buffer to receive the message
+ *  msglen    : The length of message to receive
+ *  param     :
+ *	type    : MSG_RECV_TYPE_BLOCK or MSG_RECV_TYPE_NONBLOCK\n
+ *	cb_func : user callback function for nonblocking receive
+ * 
+ * Return Value:
+ *  On success, 0 (OK) is returned.; On failure, -1 (ERROR) is returned.
+ ****************************************************************************/
+static int messaging_recv_nonblock(mqd_t *mqdes, const char *port_name, msg_recv_param_t *param)
+{
+	int ret;
+	messaging_recv_info_t *nonblock_data;
+
+	/* Check there was msg already before setting notification */
+	ret = mq_receive(*mqdes, param->msg, CONFIG_MESSAGING_UNICAST_MSGSIZE, 0);
+	if (ret > 0) {
+		(*param->cb_func)(NULL, param->msg);
+		mq_close(*mqdes);
+		mq_unlink(port_name);
```

Comment:
You should define each of situation properly.
In callback function, you calling mq_close in some situations.
In my opinion, you had better check return value and close & unlink where open mqueue(messaging_recv_internal)

Reply:
I checked all cases and add messaging_unregister API for close and unlink. Thanks.